src/VisualStudio/Core/Def/Implementation/LanguageService/AbstractPackage.cs

```diff
@@ -27,36 +27,12 @@ protected override async Task InitializeAsync(CancellationToken cancellationToke
             ForegroundObject = new ForegroundThreadAffinitizedObject(componentModel.GetService<IThreadingContext>());
         }
 
-        protected void LoadComponentsInUIContextOnceSolutionFullyLoaded(CancellationToken cancellationToken)
+        protected async Task LoadComponentsInUIContextOnceSolutionFullyLoadedAsync(CancellationToken cancellationToken)
         {
-            ThreadHelper.ThrowIfNotOnUIThread();
-
-            if (KnownUIContexts.SolutionExistsAndFullyLoadedContext.IsActive)
-            {
-                // if we are already in the right UI context, load it right away
-                LoadComponentsInUIContext(cancellationToken);
-            }
-            else
-            {
-                // load them when it is a right context.
-                KnownUIContexts.SolutionExistsAndFullyLoadedContext.UIContextChanged += OnSolutionExistsAndFullyLoadedContext;
-            }
-        }
-
-        private void OnSolutionExistsAndFullyLoadedContext(object sender, UIContextChangedEventArgs e)
-        {
-            ThreadHelper.ThrowIfNotOnUIThread();
-
-            if (e.Activated)
-            {
-                // unsubscribe from it
-                KnownUIContexts.SolutionExistsAndFullyLoadedContext.UIContextChanged -= OnSolutionExistsAndFullyLoadedContext;
-
-                // load components
-                LoadComponentsInUIContext(CancellationToken.None);
-            }
+            await KnownUIContexts.SolutionExistsAndFullyLoadedContext;
+            await LoadComponentsInUIContextAsync(cancellationToken);
         }
 
-        protected abstract void LoadComponentsInUIContext(CancellationToken cancellationToken);
+        protected abstract Task LoadComponentsInUIContextAsync(CancellationToken cancellationToken);
```

Comment:
💡 Consider renaming this to `LoadComponentsAsync` or `LoadComponentAfterSolutionLoadedAsync` to avoid readers thinking that `InUIContext` means "main thread".